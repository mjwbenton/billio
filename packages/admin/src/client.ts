import { ApolloClient, InMemoryCache } from "@apollo/client";

const CLIENT = new ApolloClient({
  uri: "http://localhost:4000/",
  cache: new InMemoryCache(),
  defaultOptions: {
    query: {
      fetchPolicy: "network-only",
    },
  },
});

export default CLIENT;
