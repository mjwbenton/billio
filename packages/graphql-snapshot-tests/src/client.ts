import { ApolloClient, InMemoryCache, HttpLink } from "@apollo/client";
import fetch from "cross-fetch";

const ENDPOINT = process.env.BILLIO_ENDPOINT;

if (!ENDPOINT) {
  throw new Error("Please configure BILLIO_ENDPOINT for tests");
}

const CLIENT = new ApolloClient({
  cache: new InMemoryCache(),
  link: new HttpLink({
    uri: ENDPOINT,
    fetch,
  }),
});

export default CLIENT;
