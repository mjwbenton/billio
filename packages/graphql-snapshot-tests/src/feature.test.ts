import client from "./client";
import { gql } from "@apollo/client";

const ITEM_MATCHER = {
  id: expect.any(String),
};

let ADDED_ID: string = "";
let IMPORTED_ID: string = "";

test("can add a feature", async () => {
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_AddFeature {
        addFeature(
          item: {
            title: "Test Feature"
            releaseYear: "2021"
            shelfId: Watched
            rating: 1
          }
        ) {
          id
          title
          releaseYear
          shelf {
            name
          }
          rating
          rewatch
        }
      }
    `,
  });
  expect(data).toMatchSnapshot({
    addFeature: ITEM_MATCHER,
  });
  ADDED_ID = data.addFeature.id;
});

test("can import external feature", async () => {
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_ImportFeature {
        importExternalFeature(externalId: "tmdb:153", shelfId: Watched) {
          id
          externalId
          title
          releaseYear
          shelf {
            id
          }
          rewatch
        }
      }
    `,
  });
  expect(data).toMatchSnapshot({
    importExternalFeature: ITEM_MATCHER,
  });
  IMPORTED_ID = data.importExternalFeature.id;
});

test("can query single feature", async () => {
  const { data } = await client.query({
    query: gql`
      query Test_QuerySingleFeature($id: ID!) {
        feature(id: $id) {
          id
          title
          releaseYear
          shelf {
            id
            name
          }
          rating
          rewatch
        }
      }
    `,
    variables: {
      id: ADDED_ID,
    },
  });
  expect(data).toMatchSnapshot({
    feature: ITEM_MATCHER,
  });
});

test("can fetch second page of features", async () => {
  const query = gql`
    query Test_FeaturePagination($after: ID) {
      features(first: 1, after: $after) {
        items {
          title
        }
        hasNextPage
        nextPageCursor
      }
    }
  `;

  const { data: first } = await client.query({
    query,
  });
  expect(first).toMatchSnapshot({
    features: {
      nextPageCursor: expect.any(String),
      hasNextPage: true,
    },
  });

  const { data: second } = await client.query({
    query,
    variables: {
      after: first.features.nextPageCursor,
    },
  });
  expect(second).toMatchSnapshot({
    features: {
      nextPageCursor: null,
      hasNextPage: false,
    },
  });
});

test("can search for external features", async () => {
  const { data } = await client.query({
    query: gql`
      {
        searchExternalFeature(term: "Lost in Translation") {
          id
          title
        }
      }
    `,
  });
  expect(data.searchExternalFeature.length).toBeGreaterThan(0);
  data.searchExternalFeature.forEach((result: unknown) => {
    expect(result).toMatchSnapshot({
      id: expect.stringMatching(/^tmdb:/),
      title: expect.any(String),
    });
  });
});

test("can mutate title on feature without change to other fields", async () => {
  const updatedTitle = "Test Feature 2";
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_MutateTitle($id: ID!, $title: String!) {
        updateFeature(id: $id, item: { title: $title }) {
          title
          releaseYear
          rating
          externalId
          notes
          image {
            url
            width
            height
          }
        }
      }
    `,
    variables: {
      id: ADDED_ID,
      title: updatedTitle,
    },
  });
  expect(data).toMatchSnapshot();
});

test("movedAt doesn't change on rating", async () => {
  const {
    data: {
      feature: { movedAt },
    },
  } = await client.query<any>({
    query: gql`
      query Test_FetchMovedAt($id: ID!) {
        feature(id: $id) {
          id
          movedAt
        }
      }
    `,
    variables: { id: ADDED_ID },
  });
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_MutateRating($id: ID!) {
        updateFeature(id: $id, item: { rating: 10 }) {
          movedAt
          rating
        }
      }
    `,
    variables: { id: ADDED_ID },
  });
  expect(data.updateFeature.rating).toEqual(10);
  expect(data.updateFeature.movedAt).toEqual(movedAt);
});

test("cannot rate feature more than 10", async () => {
  expect.assertions(1);
  try {
    await client.mutate({
      mutation: gql`
        mutation Test_InvalidRating($id: ID!) {
          updateFeature(id: $id, item: { rating: 11 }) {
            id
          }
        }
      `,
      variables: {
        id: ADDED_ID,
      },
    });
  } catch (e) {
    expect(JSON.stringify(e)).toMatch("Invalid rating: 11");
  }
});

test("can add note to a feature", async () => {
  const NOTE = "Test Note";
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_AddNote($id: ID!, $note: String!) {
        updateFeature(id: $id, item: { notes: $note }) {
          notes
        }
      }
    `,
    variables: {
      id: ADDED_ID,
      note: NOTE,
    },
  });
  expect(data.updateFeature.notes).toEqual(NOTE);
});

test("can delete features (cleanup)", async () => {
  const { data } = await client.mutate({
    mutation: gql`
      mutation Test_DeleteFeature($id1: ID!, $id2: ID!) {
        delete1: deleteFeature(id: $id1) {
          id
        }
        delete2: deleteFeature(id: $id2) {
          id
        }
      }
    `,
    variables: {
      id1: ADDED_ID,
      id2: IMPORTED_ID,
    },
  });
  expect(data.delete1.id).toEqual(ADDED_ID);
  expect(data.delete2.id).toEqual(IMPORTED_ID);
});
