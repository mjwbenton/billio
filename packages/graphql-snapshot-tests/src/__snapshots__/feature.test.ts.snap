// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add a feature 1`] = `
Object {
  "addFeature": Object {
    "__typename": "Feature",
    "id": Any<String>,
    "rating": 1,
    "releaseYear": "2021",
    "rewatch": false,
    "shelf": Object {
      "__typename": "FeatureShelf",
      "name": "Watched",
    },
    "title": "Test Feature",
  },
}
`;

exports[`can fetch second page of features 1`] = `
Object {
  "features": Object {
    "__typename": "FeaturePage",
    "hasNextPage": true,
    "items": Array [
      Object {
        "__typename": "Feature",
        "title": "Lost in Translation",
      },
    ],
    "nextPageCursor": Any<String>,
  },
}
`;

exports[`can fetch second page of features 2`] = `
Object {
  "features": Object {
    "__typename": "FeaturePage",
    "hasNextPage": false,
    "items": Array [
      Object {
        "__typename": "Feature",
        "title": "Test Feature",
      },
    ],
    "nextPageCursor": null,
  },
}
`;

exports[`can import external feature 1`] = `
Object {
  "importExternalFeature": Object {
    "__typename": "Feature",
    "externalId": "imdb:tt0335266",
    "id": Any<String>,
    "releaseYear": "2003",
    "rewatch": false,
    "shelf": Object {
      "__typename": "FeatureShelf",
      "id": "Watched",
    },
    "title": "Lost in Translation",
  },
}
`;

exports[`can mutate title on feature without change to other fields 1`] = `
Object {
  "updateFeature": Object {
    "__typename": "Feature",
    "externalId": null,
    "image": null,
    "notes": null,
    "rating": 1,
    "releaseYear": "2021",
    "title": "Test Feature 2",
  },
}
`;

exports[`can query single feature 1`] = `
Object {
  "feature": Object {
    "__typename": "Feature",
    "id": Any<String>,
    "rating": 1,
    "releaseYear": "2021",
    "rewatch": false,
    "shelf": Object {
      "__typename": "FeatureShelf",
      "id": "Watched",
      "name": "Watched",
    },
    "title": "Test Feature",
  },
}
`;

exports[`can search for external features 1`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 2`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 3`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 4`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 5`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 6`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;

exports[`can search for external features 7`] = `
Object {
  "__typename": "ExternalFeature",
  "id": StringMatching /\\^tmdb:/,
  "title": Any<String>,
}
`;
